dask/callbacks.py

```diff
@@ -136,4 +136,5 @@ def __enter__(self):
         return
 
     def __exit__(self, type, value, traceback):
-        _globals['callbacks'] = self.old
+        _globals['callbacks'].clear()
+        _globals['callbacks'].update(self.old)
```

Comment:
What if the various callbacks `__enter__` / `__exit__` calls are not nested? Can't you instead delete individual items?

Reply:
? The only way that could happen afaict is if someone is calling `__enter__`/`__exit__` manually instead of as a contextmanager. Could rewrite it to add/discard instead of copying and updating.

Reply:
> The only way that could happen afaict is if someone is calling __enter__/__exit__ manually instead of as a contextmanager. 

Or perhaps if several threads use Callbacks?

Reply:
The current handling of global state doesn't play well with threads at all. This is true for `add_callbacks` or for `set_options`. Since that would require more major changes to `_globals`, I'm apt to punt here and fix that later if at all.

Reply:
Well, it's at least an argument for adding/discarding :-)

Reply:
Done